[ADD] ec2 로의 배포 자동화와 관련된 workflow 작성

park-se-jun opened the pull request:
<!-- 이슈와 PR 이름을 맞춰주세요!! -->
<!-- 여러 라인에 걸쳐 문장이 필요한 경우, 각 라인마다 "-" 를 이용해주세요 -->

## PR Description 🗒️
<!-- 
PR에 대한 간략한 설명을 적어주세요
ex)
- 로그인 기능 추가.
-->
- cd 관련 workflow 작성

close #126 
## 추가 사항 및 설명 ➕
<!-- 
[FEAT] / [ADD]
추가 사항과 이에 대한 설명을 적어주세요
ex)
- 예매 기능 추가
    - 
-->

- application yml 의 뼈대 추가
- application.yml에 h2와 prod profile 추가
- h2 사용시 user 테이블이 예약어이기 때문에 생성되지 않는 문제를 해결하기 위해  h2 버전 업
- 민감한 정보는 .env 파일로 관리 ( .github ignore에 추가함)
- 로컬에서 테스트할 때 env 정보를 포함하여 실행시키고 싶다면 하고 싶다면 application.yml의 주석을 해재해 줘야함
- cd 관련 github secrets 추가
- 변경된 application.yml에 맞게 file.image를 local.image.path로 변경
- 서버 어플리케이션을 감싸는 dockerfile 작성
-  dockerfile 생성시 사용하기 위한 entrypoint를 shell script 로 작성
- 서버 어플리케이션을 감싸는 도커의 원격 저장소 생성(https://github.com/park-se-jun/BE-03-InterMark/pkgs/container/intermark)
- mysql 만 만들던 docker-compose 를 classpath의 root로 이동 및 서버를 감싼 도커도 포함하도록 변경
## 수정 사항 및 이유 🔄
<!-- 
[CHORE] / [FIX] / [HOTFIX] / [REFACTOR]
수정 사항과 수정한 이유를 작성해주세요.
ex)
- 파라미터 타입을 Long에서 long으로 수정.
    - null이 포함되지 않고 단순 연산만 수행하여 수정.
-->

## To Reviewers 🙏
<!-- 
리뷰어들이 중점적으로 봐줬으면 하는 부분이나 
리뷰어들에게 추가적으로 하고 싶은 말을 적어주세요!! 
-->
- 적용시키면서 사용한 env 파일은 카카오톡으로 배포하겠습니다.
- develop에서 서버로 배포된 상태는 docker-compose 명령을 통해 로컬에서 확인할 수 있습니다.
(docker-compose.yml과 같은 경로에 ./env/db.env, ./env/spring.env 가있어야함)

park-se-jun committed:
[ADD] application.yml 관련 파일 추가
- 기본적으로 필요한 공용 application.yml 들 추가.
- 민감한 정보는 외부 환경변수를 참조하도록 만듦.
- h2 database 이용시 user 가 예액어로 걸려있던 문제를 피하기 위해 h2 버전 업
- UploadImageServiceImpl.java 의 file.images를 application.yml을 만들면서 local.images.path로 변경

park-se-jun committed:
[ADD] 배포관련 docker 파일들 작성
- 현재 빌드 결과를 감싸는 dockerfile 생성
- 현재 파일 + mysql을 연결하는 docker-compose.yml 연결
- dockerfile의 Entrypoint를 스크립트 명령어로 작성

park-se-jun committed:
[ADD] : 배포관련 github actions workflow 작성 및 docker-compose 수정
- workflow에 프로젝트를 빌드 하는 로직 추가
- 빌드된 프로젝트를 기준으로, 감싼 docker 를 원격 저장소로 push
- scp 를 이용해 프로젝트의 docker-compose.yml 을 ec2로 복사
- ec2에서 원격 저장소의 docker image를 받아와 빌드할 수 있도록 docker-compose.yml 수정
- docker-compose.yml에서 사용하는 env 파일들을 secrets 로부터 ec2에 생성
- ec2에 이미 생성된 compose 가 있을 경우 db를 빼고 삭제했다가, 새로 빌드 하도록 만듦

park-se-jun committed:
[FIX] : 오타로 인해 수정이 안되는 것 반영

- run-on을 runs-on으로 변경

park-se-jun committed:
[FIX] : 오타 수정

- docker/login-actions@v2 에서 docker/login-action@v2로 변경

park-se-jun committed:
[ADD] : 이미지에 latest 태그 붙이는 방법을 올바른 방법으로 변경
- tag를 붙이고 push 하도록 변경

park-se-jun committed:
[FIX] : 지역환경 변수 제거

- 지역환경 변수가 동작하지 않기 때문에 이를 제거함

park-se-jun committed:
[FIX] : git describe --tags 대신 GITHUB_SHA 사용


- 태그지정을 위해 GITHUB_SHA를 대체 사용
- 아직 제거되지 않은 지역 환경변수 관련 구문 제거

park-se-jun committed:
[FIX] docker Image tag 설졍 변경

park-se-jun committed:
[FIX] push docker image 스탭에도 지역환경변수설정

park-se-jun committed:
[FIX] : deploy 관련 step을 수정
- scp로 docker-compose를 copy 할 때 overwrite 하도록 수정
- ssh접속 후 env 파일을 만드는데 echo 가 누락된 것을 수정

park-se-jun committed:
[FIX] deploy 과정에서 env 가 없을 경우에만 mkdir 이 동작하도록 설정

park-se-jun committed:
[FIX] echo ** > ** 가 env를 생성하지 못하는 문제

- 환경변수를 큰 따옴표로 감싸 주었음

park-se-jun committed:
[ADD] : docker image를 최신으로 유지하기 위한 명령어 추가

- docker-compose pull 을 docker-compose-up 전에 추가

park-se-jun committed:
[DEL] develop에 push 했을 경우만 동작하도록 변경

- on.push.branch 에서 현재 브랜치의 이름(ADD/#126) 제거
[skip ci]

Seongjun-Kwon commented on src/main/resources/application.yml:
```diff
@@ -0,0 +1,62 @@
+        name: ${BUCKET_NAME}
+        url: ${BUCKET_URL}
+    region:
+      static: ap-northeast-요2
```
ap-northeast-2 로 수정해주세용

park-se-jun replied:
변경했씁니다

Seongjun-Kwon reviewed (approved):
고생하셨습니다~

twotwobread reviewed (approved):
수고하셨습니다!!

thguss commented on .github/workflows/cd-to-ec2.yml:
```diff
@@ -0,0 +1,76 @@
+        uses: docker/login-action@v2
+        with:
+          registry: ghcr.io
+          username: ${{secrets.GHCR_USERNAME}}
```
이 부분은 깃허브 액션 secret 키에 등록되어 있는 정보인가요~?~?

park-se-jun replied:
넵 setting에서 secret and variable - action으로 들어가면 어떤 변수들이 secrets에 등록되어 있는지 확인할 수 있습니다.

thguss reviewed (approved):
수고하셨습니다~!

park-se-jun committed:
[FIX] cloud.region.static 이 잘못되었던거 수정

- ap-northeast-요2 로 잘못 써져 있었음

park-se-jun committed:
[ADD] application.yml 수정

- 여러 브랜치에서 지속적으로 file.images 관련 이슈들이 발생한 관계로 해당 config 를 추가
- 만약 로컬환경에서 .env 를 포함시키고 싶을때 .env 파일의 결로를 프로젝트의 최상위 경로가 아닌 /env/spring.env 에 위치하면 해당 env 를 읽어오도록 주석 내용을 변경
- 특정 브랜치에서 JPA repository가 생성될 때 Access to DialectResolutionInfo cannot be null when 'hibernate.dialect' not set 에러가 발생하는 관계로 hibernate의 dialect를 설정해주는 database-platform 추가
- cloud.aws.credentials 의 access-key와 secret-key를 카멜케이스로 변경

park-se-jun committed:
[ADD] docker-compose설정

- ec2 가 reboot 되도 실행 되도록 설정
- 각각의 serice에 timezone 을 seoul로 설정

park-se-jun committed:
[ADD] develop과의 conflict 해서

- merge develop 을 통해 현재 develop과의 conflict 해소

park-se-jun committed:
[FIX] : develop 과  conflict 해소
